arrow-fx-coroutines/README.MD

```diff
@@ -0,0 +1,480 @@
+   Resource({ openFile("path3") }, File::close)
+ )
+
+val resource: Resource<List<File>> =
+  resources.sequence(Resource.applicative())
+
+suspend main(): Unit {
+  resource.use { files ->
+    files.parTraverse(IODispatchers.IOPool) { file ->
+       file.toString()
+    }
+  }
+}
+```
+
+### Error Handling
+
+In Kotlin with suspend `try/catch` can safely be used to recover from exceptions.
+
+Simple constructs like `suspend fun Either.catch(f: () -> A): Either<Throwable, A>` are available for nicer syntax when and lifting errors in your domain.
+
+### Retrying and repeating effects
+
+`Schedule` allows you to define and compose powerful yet simple policies, which can be used to either repeat or retry computation. 
+
+A simple example might be to repeat an action `n` times, similar to the `repeat` function in the standard library.
+
+```kotlin:ank
+suspend main(): Unit {
+  repeat(Schedule.recurs<A>(n)) {
+    println("Hello")
+  }
+}
+```
+
+Alternatively we can re-use this `Schedule` to `retry` a `suspend fun` `n` times when it fails.
+
+```kotlin:ank
+suspend main(): Unit {
+  retry(Schedule.recurs<A>(n)) {
+    println("I am going to do nothing but throw a tantrum!")
+    throw RuntimeException("Boom!")
+  }
+}
+```
+
+Additionally a `Schedule` can also produce an output, let's compose our simple `recurs` Schedule to `recur` every `10.seconds` in a `spaced` fashion and collect all outputs along the way.
+
+```kotlin:ank
+fun <A> schedule(): Schedule<A, List<A>> = Schedule {
+  (recurs<A>(10) and spaced(10.seconds)) zipRight collect()
+}
+
+suspend main(): Unit {
+  var count = Atomic(0)
+
+  val history: List<Int> = repeat(schedule<Int>()) {
+    println("Incrementing the ref")
+    count.update(Int::inc)
+  }
+}
+```
+
+## Parallel Ops
+
+
+### ForkConnected
+
+`forkConnected` launches a `suspend () -> A` function in a seperate processes called a `Fiber`, the lifecycle of this process is connected to its parent.
+If the parent cancels that this operation will also be cancelled.
+
+A `Fiber` is a data type that represents a `suspend fun join(): A` and `suspend fun cancel(): Unit` handle.
+
+### ForkScoped
+
+`forkScoped` is a similar construct as `ForkConnected` but instead of automatically connecting to its parent it takes a `suspend () -> Unit` op which triggers the cancellation.
+This can be used as a lower level construct to signal cancellation on a trigger, and wire cancellation with 3rd party frameworks.
+
+```kotlin:ank
+tailrec fun fun sleeper(): Unit {
+  sleep(1.seconds)
+  return sleeper()
+}
+
+
+class AndroidFragnent {
+
+  val cancelPromise = Promise.unsafe<Unit>()
+  
+  override fun onCreateView(bundle: SavedBundleInstance): Unit {
+     suspend {
+       ::sleeper.forkScoped(cancelPromise::get)
+     }.startCoroutine(Continuation(EmptyCoroutineContext) { })
+  }
+  
+  override fun onDestroyView(): Unit {
+    suspend { cancelPromise.complete(Unit) }
+      .startCoroutine(Continuation(EmptyCoroutineContext) { })
```

Comment:
Should we provide an API that subsumes this pattern?

```
inline fun startWithContinuation(c: Continuation, f: suspend () -> A) =
  suspend { f() }
      .startCoroutine(c)
```
```

Reply:
Yes, we should but I haven't figured out the best solution here.

An `Enviroment` interface could offer a nice API for this.